feat: testing guidelines (#377)
nodejs/package-maintenance: docs/drafts/testing-guidelines.md

* LTS (long time support)
* Current

> Of course, you are free to maintain a package that also runs with older versions of Node.js that have reached the "end-of-life" stage.

Reviewer: Perhaps something a bit stronger than "free to"? I'd personally prefer "encouraged to".

Author: agreed.

> Of course, you are free to maintain a package that also runs with older versions of Node.js that have reached the "end-of-life" stage.

Reviewer: ```suggestion
Of course, you are encouraged to maintain a package that also runs with older versions of Node.js that have reached the "end-of-life" stage.
```

Reviewer 2: I think "encouraged to maintain" comes across as too strong in that sentence, and suggests an expansive maintenance policy rather than an appropriate testing policy. The context here is what versions are being tested, and testing a range? So perhaps:

Of course, you are encouraged to also test against the older versions of Node.js your package supports.

Reviewer: I'm suggesting that the expansive policy is the appropriate one. Only covering LTS is a minimum, not the ideal.

Reviewer 2: I'll give a long reply since the expansive policy was deliberate, and caught my attention as something I question as appropriate in the Testing Guidelines context.

I think expansive is overly aspirational for maintainers. Encouraging maintainers to support a version of node that nodejs itself does not support, and will mean their package is running on an unsupported platform not getting bugs fixes or security fixes, likely with dependencies that are not getting bugs fixes or security fixes (from personal experience), seems a compromised version of "maintain" to encourage.

I did a quick search in case I was being overly pessimistic, but my first search results was:
https://medium.com/the-node-js-collection/maintainers-should-consider-following-node-js-release-schedule-ab08ed4de71a

I think coverage and pros and cons of support policies is an interesting separate topic, and don't like encouraging a particular pattern embedded in the Testing recommendations. 

(My impression from some comments in other topics is that it would be an improvement for new versions of node if more packages tested current and lts consistently? Impression from following limited number of the drafts and discussion, no evidence to quote!)

Not sure if others may agree, but passing on my comments. Thanks!

Reviewer 3: I am +1 to encouraging (heh) folks to stick with Node.js Maintenance releases being the furthest back release line we push projects to support. I agree that we shouldn't expect folks to support versions that we ourselves don't support. Doing so does also help move the ecosystem forward both in terms of folks being able to consistently use new language features *and* in terms of Node.js itself being able to move forward more easily.

Reviewer: My belief differs; supporting an older version of node helps make those users' *upgrade of node itself* easier.

I'm fine with these docs not pushing for supporting EOL node versions, but I am not fine with them *discouraging* doing so in any way, including implying that it's not a big deal.

Author: So lots of discussion on this point. I think that the wording **encouraged to** leaves this open to the maintainer while indicating that support of more versions is better. I believe that we will not be able to resolve this insomuch as giving definitive advice because both sides of this discussion have valid fundamentals but differ. 

I would like to suggest that we compromise but from previous discussions on points of compromise it has been pointed out that we are actually supposed to be making **clear recommendations**.

So what would be a no compromise settlement?

I think if we say that support of LTS is considered the minimal but encourage as much support as is possible may satisfy both sides.

Thoughts?

Reviewer: I'm very content with that.

Reviewer 2: Short version: What about "Supporting older versions of node is encouraged where practical."?

-----

Long version

Thanks @ghinks, and I appreciate the explanation of trying for a **clear recommendation** rather than compromising early.

If "as much support as is possible" is the intended wording then my concerns still apply, that not only does this encourage a policy which is somewhat contentious, but also is not realistic for many maintainers. I am personally more concerned with the second issue, so exploring that...

A long suggestion, somewhat tongue in cheek:
"Supporting older versions of node is encouraged for the convenience of your package consumers who run older versions of node, but may not be practical due to restricting what platform features you can use, the support policies of your own dependencies, or the overheads of maintaining multiple streams of development."

This collapses down to something which might be acceptable to more people:
"Supporting older versions of node is encouraged where practical.".

The main issue that caused me to change my own support policies to "fully supported on Long Term Support versions of Node" was the testing frameworks. Not all packages state their support policy (which will improve as [PACKAGE-SUPPORT](https://github.com/nodejs/package-maintenance/blob/master/docs/PACKAGE-SUPPORT.md) is adopted, but looking at `package.json` for two popular ones I have used, currently:

Jest:  `"node": ">= 10.14.2"`

Mocha: `"node": ">= 10.12.0"`

I did also find explicit info in the Mocha [PROJECT_CHARTER](https://github.com/mochajs/mocha/blob/master/PROJECT_CHARTER.md) where it states that "Use within unmaintained versions of Node.js" is "Out of Scope".

Reviewer: I'm also fine with the "where practical" qualifier, since in my experience it's impractical for very few projects to actually preserve support for very old node versions :-)

You're totally right though about testing frameworks; the only one I'm aware of that won't ever force you to drop support for some of your users is tape.

Reviewer 3: I am fine with either variant proposed.

> For browser based modules and modules that are designed to work equally both client and server side it is essential to test on various 

Reviewer: ```suggestion
For browser-based modules and modules that are designed to work equally both client and server side it is essential to test on various 
```

* **integration test**: test your code with other applications dependencies
* **acceptance test**: test your application sticks in performance, heavy load, etc..

> #### Unit Testing

Reviewer 4: Should we not just link to something like https://martinfowler.com/bliki/TestPyramid.html?

Author: I am happy to add a link

Reviewer 4: Sorry, wasn't sure - did you mean to do it in this PR or later?

Author: I will add a link later. If that is ok. This document is going to be revisited after I do the CI/CD document that this document will link to aswell.

The initial and fundamental start of any testing strategy should start with a unit testing strategy. For the JavaScript language
the unit test may be the first place to catch code that will not parse. It is also the way to define the expectation for a particular
function or code block. The unit test should describe what to expect from that code when given a defined set of inputs.

For a test to be considered a unit test it should consume no remote services.

It is generally considered good practice to run units as part of any commit or build strategy.

#### Integration Testing
Integration testing in general requires that the code under test be in a package.

For many packages integration testing requires that the built package be run via a testing tool upon several environments. The /
field of continuous integration and deployment (CICD) is complex. The package being built shall be run through

> a CICD pipeline. Several integrations are available for GitHub based repositories. (There are also alternatives to GitHub.)

Reviewer 4: Pet peeve, but CI/CD pipeline is so incredibly vague, that it can mean pretty much anything... While I'm not sure we need to include these details at all (nor that the definitions of unit vs integration matter that much at all, fwiw), I think we should still link to some examples or in depth articles on this.

I'd very much be in favor of how this is done in JS world, though! Possibly with a distinction on how you do certain things in node vs browser.

Author: Agreed. We even have the whole document on [CI/CD](https://github.com/nodejs/package-maintenance/blob/master/docs/drafts/ci-cd-guidelines.md) dedicated to this subject. 

For the testing document maybe we should reference our CI/CD draft. 

What both yourself and @bnb have contributed to this document too.

I think we do need to have CI/CD Options documented.
- what can we do for free for an open source project
- what providers are there? 
- what does each give us?

This is actually a pretty big area in both open source and the enterprise and I think we should talk about this as an item on 14th July in the Package Maintenance meeting.

Author: @dominykas are we ok referring to the CI/CD document. I agree that we need to give concrete examples of what is currently possible in that document and I intend to start that document next.

Reviewer 4: Sure, we can link and extend the existing doc. I'm just not entirely sure about the term "CI/CD pipeline" - it can mean a gazillion things to different people, so possibly we should simplify the wording towards something like "your tests should run automatically on pr/merge/commit/etc" and link to the doc?

It should not matter if it's a "pipeline" or a single job on Travis. It does not need to be complex. People need not be scared of testing and automation. The simpler the better, esp in the context of oss node packages.

Author: agreed.

* **integration test**: test your code with other applications dependencies
* **acceptance test**: test your application sticks in performance, heavy load, etc..

#### Unit Testing
The initial and fundamental start of any testing strategy should start with a unit testing strategy. For the JavaScript language

> the unit test may be the first place to catch code that will not parse. It is also the way to define the expectation for a particular

Reviewer: ```suggestion
the unit test may be the first place to catch code that will not parse (a linter can achieve this as well). It is also the way to define the expectation for a particular
```

function or code block. The unit test should describe what to expect from that code when given a defined set of inputs.

> For a test to be considered a unit test it should consume no external services and only the code block under test should be considered part of that test. All external services that the code block under test consumes should be

Reviewer: > only the code block under test should be considered part of that test

I don't agree with this; see https://martinfowler.com/bliki/UnitTest.html - unit tests should imo avoid mocking, and should effectively be an integration test of the intended behavior.

Author: Agreed, this can be an area of contention. If I make the text clearer to indicate that the intent is to avoid the use of services such as DBs and active end points I believe it would clarify the intent I had here.

Author: Altered accordingly.

stubbed. There are many mocking tools available and this document will not favor one set of tools over any other.

It is generally considered good practice to run units as part of any commit or build strategy.

#### Integration Testing
Integration testing in general requires that the code under test be in a package.

For many packages integration testing requires that the built package be run via a testing tool upon several environments. The 

> field of continuous integration and deployment (CICD) is complex. The package being built shall be run through

Reviewer: ```suggestion
field of continuous integration and deployment (CI/CD) is complex. The package being built shall be run through
```

> a CICD pipeline. Several integrations are available for GitHub based repositories. (There are also alternatives to GitHub.)

Reviewer: ```suggestion
a CI/CD pipeline. Several integrations are available for GitHub based repositories. (There are also alternatives to GitHub.)
```
